src/alternatives.js

```diff
@@ -0,0 +1,136 @@
+import inherits from './util/inherits';
+import MixedSchema from './mixed';
+import { alternatives as locale } from './locale.js';
+import isSchema from './util/isSchema';
+import typeOf from 'type-name';
+import { SynchronousPromise } from 'synchronous-promise';
+import createValidation from './util/createValidation';
+import printValue from './util/printValue';
+import runValidations from './util/runValidations';
+
+let promise = sync => sync ? SynchronousPromise: Promise;
+const invalidMarker = {};
+
+
+export default function AlternativesSchema() {
+    if (!(this instanceof AlternativesSchema))
+        return new AlternativesSchema()
+
+    MixedSchema.call(this, { type: 'alternatives' })
+}
+
+inherits(AlternativesSchema, MixedSchema, {
+
+    _validate(_value, options = {}) {
+        let value  = _value;
+        let originalValue = options.originalValue != null ?
+            options.originalValue : _value
+
+        let isStrict = this._option('strict', options)
+        let endEarly = this._option('abortEarly', options)
+
+        let sync = options.sync
+        let path = options.path
+        let label = this._label
+
+        if (!isStrict) {
+            value = this._cast(value, { assert: false, ...options })
+        }
+        // value is cast, we can check if it meets type requirements
+        let validationParams = { value, path, schema: this, options, label, originalValue, sync }
+        let initialTests = []
+
+        if (this._oneOfType)
+            initialTests.push(this._oneOfType(validationParams));
+
+        return runValidations({ validations: initialTests, endEarly, value, path, sync })
+        .then(value => runValidations({
+            path,
+            sync,
+            value,
+            endEarly,
+            validations: this.tests.map(fn => fn(validationParams)),
+        }))
+    },
+
+    resolveSchemaUsingValue(value){
+        if(this._oneOfTypeOptions){
```

Comment:
this does work but will break if there are any async validations which isn't ideal...Maybe it's enough to just use cast like you use validate for the validate case e.g.

```
for(let schema of this._oneOfTypeOptions) {
  try { 
     return schema.cast(value) 
  } catch (err) { 
     continue //failed to cast
  }
}
```

Reply:
I like this solution, will change

Reply:
What should happen when a cast is attempted on a value which is not supported? Throw error?

Reply:
yeah that's what this does: https://github.com/jquense/yup/blob/master/src/mixed.js#L152